File: app/src/main/java/org/wikipedia/feed/announcement/AnnouncementCard.java
package org.wikipedia.feed.announcement;

import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.wikipedia.feed.model.Card;
import org.wikipedia.feed.model.CardType;

public class AnnouncementCard extends Card {
    @NonNull private final Announcement announcement;

    public AnnouncementCard(@NonNull Announcement announcement) {
        this.announcement = announcement;
    }

    @Override @NonNull public String title() {
        return announcement.type();
    }

    @Nullable @Override public String extract() {
        return announcement.text();
    }

    public boolean hasAction() {
        return announcement.hasAction();
    }

    @NonNull public String actionTitle() {
        return announcement.actionTitle();
    }

    @NonNull public Uri actionUri() {
        return Uri.parse(announcement.actionUrl());
    }

    @Nullable public String negativeText() {
        return announcement.negativeText();
    }

    public boolean hasFooterCaption() {
        return announcement.hasFooterCaption();
    }

    @NonNull public String footerCaption() {
        return announcement.footerCaption();
    }

    public boolean hasImage() {
        return announcement.hasImageUrl();
    }

    boolean hasBorder() {
        return announcement.hasBorder();
    }

    @NonNull public boolean isArticlePlacement() {
        return Announcement.PLACEMENT_ARTICLE.equals(announcement.placement());
    }

    @NonNull @Override public Uri image() {
        return Uri.parse(announcement.imageUrl());
    }

    public int imageHeight() {
        try {
            return Integer.parseInt(announcement.imageHeight());
        } catch (NumberFormatException e) {
            return 0;
        }
    }

    @NonNull @Override public CardType type() {
        return CardType.ANNOUNCEMENT;
    }

    @Override protected int dismissHashCode() {
        return announcement.id().hashCode();
    }
}

File: app/src/main/java/org/wikipedia/feed/announcement/AnnouncementCard.kt
package org.wikipedia.feed.announcement

import android.net.Uri
import org.wikipedia.feed.model.Card
import org.wikipedia.feed.model.CardType

open class AnnouncementCard(private val announcement: Announcement) : Card() {

    val isArticlePlacement get() = Announcement.PLACEMENT_ARTICLE == announcement.placement

    override fun title(): String {
        return announcement.type
    }

    override fun extract(): String? {
        return announcement.text
    }

    override fun image(): Uri {
        return Uri.parse(announcement.imageUrl.orEmpty())
    }

    override fun type(): CardType {
        return CardType.ANNOUNCEMENT
    }

    override fun dismissHashCode(): Int {
        return announcement.id.hashCode()
    }

    fun imageHeight(): Int {
        return try {
            announcement.imageHeight.orEmpty().ifEmpty { "0" }.toInt()
        } catch (e: NumberFormatException) {
            0
        }
    }

    fun hasAction(): Boolean {
        return announcement.hasAction()
    }

    fun actionTitle(): String {
        return announcement.actionTitle()
    }

    fun actionUri(): Uri {
        return Uri.parse(announcement.actionUrl())
    }

    fun negativeText(): String? {
        return announcement.negativeText
    }

    fun hasFooterCaption(): Boolean {
        return announcement.hasFooterCaption()
    }

    fun footerCaption(): String {
        return announcement.footerCaption.orEmpty()
    }

    fun hasImage(): Boolean {
        return announcement.hasImageUrl()
    }

    fun hasBorder(): Boolean {
        return announcement.border == true
    }
}
